File: src/main/scala/ff/core/Option.scala
package ff.core
import ff.core.Array_._

import ff.core.ArrayBuilder_._

import ff.core.Bool_._

import ff.core.Char_._

import ff.core.Core_._

import ff.core.FileSystem_._

import ff.core.Int_._

import ff.core.List_._

import ff.core.Log_._

import ff.core.Map_._

import ff.core.Nothing_._

import ff.core.Option_._

import ff.core.Pair_._

import ff.core.Set_._

import ff.core.String_._

import ff.core.System_._

import ff.core.Try_._

import ff.core.Unit_._
object Option_ {

type Option[+T] = scala.Option[T];

object None {
def apply[T]() = scala.None;
def unapply[T](value : Option[T]) = value == scala.None;
}
object Some {
def apply[T](value_ : T) = scala.Some(value_ : T);
def unapply[T](value : Option[T]) = scala.Some(value).collectFirst { case scala.Some(value_ : T) => (value_ : T) };
}


def Option_else[T, S >: T, T2 <: S](self_ : Option[T], body_ : Function0[T2]) : S = { pipe_dot(self_)({ case (ff.core.Option_.None()) => body_(); case (ff.core.Option_.Some(value_)) => value_})}

def Option_elseIf[T, S >: T, T2 <: S](self_ : Option[T], condition_ : Function0[ff.core.Bool_.Bool], body_ : Function0[T2]) : ff.core.Option_.Option[S] = { pipe_dot(self_)({ case (ff.core.Option_.None()) => pipe_dot(condition_())({ case (ff.core.Bool_.True()) => ff.core.Option_.Some(body_()); case (ff.core.Bool_.False()) => ff.core.Option_.None()}); case (ff.core.Option_.Some(_)) => self_ })}

def Option_getElse[T, S >: T, T2 <: S](self_ : Option[T], body_ : Function0[ff.core.Option_.Option[T2]]) : ff.core.Option_.Option[S] = { pipe_dot(self_)({ case (ff.core.Option_.None()) => body_(); case (ff.core.Option_.Some(_)) => self_ })}

def Option_getEmpty[T](self_ : ff.core.Option_.Option[T]) : ff.core.Bool_.Bool = (self_) match {
case (self_) =>
pipe_dot(self_)({
case (ff.core.Option_.None()) =>
ff.core.Bool_.True()
case (ff.core.Option_.Some(_)) =>
ff.core.Bool_.False()
})
}

def Option_getList[T](self_ : ff.core.Option_.Option[T]) : ff.core.List_.List[T] = (self_) match {
case (self_) =>
pipe_dot(self_)({
case (ff.core.Option_.None()) =>
List()
case (ff.core.Option_.Some(v_)) =>
List(v_)
})
}

def Option_filter[T](self_ : ff.core.Option_.Option[T], body_ : Function1[T, ff.core.Bool_.Bool]) : ff.core.Option_.Option[T] = (self_, body_) match {
case (self_, _) =>
pipe_dot(self_)({
case (ff.core.Option_.Some(v_)) if body_(v_) =>
ff.core.Option_.Some[T](value_ = v_)
case (_) =>
ff.core.Option_.None[T]()
})
}

def Option_map[T, R](self_ : ff.core.Option_.Option[T], body_ : Function1[T, R]) : ff.core.Option_.Option[R] = (self_, body_) match {
case (self_, _) =>
pipe_dot(self_)({
case (ff.core.Option_.None()) =>
ff.core.Option_.None[R]()
case (ff.core.Option_.Some(v_)) =>
ff.core.Option_.Some[R](value_ = body_(v_))
})
}

def Option_flatMap[T, R](self_ : ff.core.Option_.Option[T], body_ : Function1[T, ff.core.Option_.Option[R]]) : ff.core.Option_.Option[R] = (self_, body_) match {
case (self_, _) =>
pipe_dot(self_)({
case (ff.core.Option_.None()) =>
ff.core.Option_.None[R]()
case (ff.core.Option_.Some(v_)) =>
body_(v_)
})
}

def Option_each[T](self_ : ff.core.Option_.Option[T], body_ : Function1[T, ff.core.Unit_.Unit]) : ff.core.Unit_.Unit = (self_, body_) match {
case (self_, _) =>
pipe_dot(self_)({
case (ff.core.Option_.None()) =>
ff.core.Unit_.Unit()
case (ff.core.Option_.Some(v_)) =>
body_(v_)
});
ff.core.Unit_.Unit()
}

def Option_all[T](self_ : ff.core.Option_.Option[T], body_ : Function1[T, ff.core.Bool_.Bool]) : ff.core.Bool_.Bool = (self_, body_) match {
case (self_, _) =>
pipe_dot(self_)({
case (ff.core.Option_.None()) =>
ff.core.Bool_.True()
case (ff.core.Option_.Some(v_)) =>
body_(v_)
})
}

def Option_any[T](self_ : ff.core.Option_.Option[T], body_ : Function1[T, ff.core.Bool_.Bool]) : ff.core.Bool_.Bool = (self_, body_) match {
case (self_, _) =>
pipe_dot(self_)({
case (ff.core.Option_.None()) =>
ff.core.Bool_.False()
case (ff.core.Option_.Some(v_)) =>
body_(v_)
})
}

def Option_expect[T](self_ : ff.core.Option_.Option[T]) : T = (self_) match {
case (self_) =>
pipe_dot(self_)({
case (ff.core.Option_.None()) =>
ff.core.Core_.panic_[T](message_ = "None.expect()")
case (ff.core.Option_.Some(v_)) =>
v_
})
}

def Option_contains[T](self_ : ff.core.Option_.Option[T], value_ : T) : ff.core.Bool_.Bool = (self_, value_) match {
case (self_, _) =>
pipe_dot(self_)({
case (ff.core.Option_.None()) =>
ff.core.Bool_.False()
case (ff.core.Option_.Some(v_)) =>
(v_ == value_)
})
}

def Option_flatten[T](self_ : ff.core.Option_.Option[ff.core.Option_.Option[T]]) : ff.core.Option_.Option[T] = (self_) match {
case (self_) =>
pipe_dot(self_)({
case (ff.core.Option_.None()) =>
ff.core.Option_.None[T]()
case (ff.core.Option_.Some(v_)) =>
v_
})
}


}

File: src/main/js/ff/core/Option.mjs
import * as ff_core_Array from "../../ff/core/Array.mjs"

import * as ff_core_ArrayBuilder from "../../ff/core/ArrayBuilder.mjs"

import * as ff_core_Bool from "../../ff/core/Bool.mjs"

import * as ff_core_Char from "../../ff/core/Char.mjs"

import * as ff_core_Core from "../../ff/core/Core.mjs"

import * as ff_core_FileSystem from "../../ff/core/FileSystem.mjs"

import * as ff_core_Int from "../../ff/core/Int.mjs"

import * as ff_core_List from "../../ff/core/List.mjs"

import * as ff_core_Log from "../../ff/core/Log.mjs"

import * as ff_core_Map from "../../ff/core/Map.mjs"

import * as ff_core_Nothing from "../../ff/core/Nothing.mjs"

import * as ff_core_Option from "../../ff/core/Option.mjs"

import * as ff_core_Pair from "../../ff/core/Pair.mjs"

import * as ff_core_Set from "../../ff/core/Set.mjs"

import * as ff_core_String from "../../ff/core/String.mjs"

import * as ff_core_System from "../../ff/core/System.mjs"

import * as ff_core_Try from "../../ff/core/Try.mjs"

import * as ff_core_Unit from "../../ff/core/Unit.mjs"

// type Option
const None$ = {_: 'None'};
export function None() {
return None$;
}
export function Some(value_) {
return {_: 'Some', value_};
}





export function Option_else(self_, body_) {
{
const _1 = self_
{
if(_1._ === 'None') {
return body_()
return
}
}
{
if(_1._ === 'Some') {
const value_ = _1.value_
return value_
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function Option_elseIf(self_, condition_, body_) {
{
const _1 = self_
{
if(_1._ === 'None') {
{
const _1 = condition_()
{
if(_1) {
return ff_core_Option.Some(body_())
return
}
}
{
if(!_1) {
return ff_core_Option.None()
return
}
}
throw new Error('Unexhaustive pattern match')
}
return
}
}
{
if(_1._ === 'Some') {
return self_
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function Option_orElse(self_, body_) {
{
const _1 = self_
{
if(_1._ === 'None') {
return body_()
return
}
}
{
if(_1._ === 'Some') {
return self_
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function Option_isEmpty(self_) {
{
const _1 = self_
{
if(_1._ === 'None') {
return true
return
}
}
{
if(_1._ === 'Some') {
return false
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function Option_toList(self_) {
{
const _1 = self_
{
if(_1._ === 'None') {
return ff_core_List.Empty()
return
}
}
{
if(_1._ === 'Some') {
const v_ = _1.value_
return ff_core_List.Link(v_, ff_core_List.Empty())
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function Option_filter(self_, body_) {
{
const _1 = self_
{
if(_1._ === 'Some') {
const v_ = _1.value_
if(body_(v_)) {
return ff_core_Option.Some(v_)
return
}
}
}
{
return ff_core_Option.None()
return
}
throw new Error('Unexhaustive pattern match')
}
}

export function Option_map(self_, body_) {
{
const _1 = self_
{
if(_1._ === 'None') {
return ff_core_Option.None()
return
}
}
{
if(_1._ === 'Some') {
const v_ = _1.value_
return ff_core_Option.Some(body_(v_))
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function Option_flatMap(self_, body_) {
{
const _1 = self_
{
if(_1._ === 'None') {
return ff_core_Option.None()
return
}
}
{
if(_1._ === 'Some') {
const v_ = _1.value_
return body_(v_)
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function Option_each(self_, body_) {
for(;;) {
const _1 = self_
{
if(_1._ === 'None') {

break
}
}
{
if(_1._ === 'Some') {
const v_ = _1.value_
body_(v_)
break
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function Option_all(self_, body_) {
{
const _1 = self_
{
if(_1._ === 'None') {
return true
return
}
}
{
if(_1._ === 'Some') {
const v_ = _1.value_
return body_(v_)
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function Option_any(self_, body_) {
{
const _1 = self_
{
if(_1._ === 'None') {
return false
return
}
}
{
if(_1._ === 'Some') {
const v_ = _1.value_
return body_(v_)
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function Option_expect(self_) {
{
const _1 = self_
{
if(_1._ === 'None') {
return ff_core_Core.panic_("None.expect()")
return
}
}
{
if(_1._ === 'Some') {
const v_ = _1.value_
return v_
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function Option_contains(self_, value_) {
{
const _1 = self_
{
if(_1._ === 'None') {
return false
return
}
}
{
if(_1._ === 'Some') {
const v_ = _1.value_
return (v_ == value_)
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function Option_flatten(self_) {
{
const _1 = self_
{
if(_1._ === 'None') {
return ff_core_Option.None()
return
}
}
{
if(_1._ === 'Some') {
const v_ = _1.value_
return v_
return
}
}
throw new Error('Unexhaustive pattern match')
}
}


